src/bios.rs

```diff
@@ -0,0 +1,115 @@
+use std::io::prelude::*;
+use std::path::Path;
+
+use crate::component::*;
+use crate::model::*;
+use anyhow::{bail, Result};
+use crate::ostreeutil;
+use crate::util;
+
+// grub2-install file path
+pub(crate) const GRUB_BIN: &str = "usr/sbin/grub2-install";
+
+#[allow(dead_code)]
+#[derive(Default)]
+pub(crate) struct Bios {}
+
+impl Component for Bios {
+    fn name(&self) -> &'static str {
+        "BIOS"
+    }
+
+    fn install(&self, src_root: &openat::Dir, dest_root: &str, device: &str) -> Result<InstalledContent> {
+        let meta = if let Some(meta) = get_component_update(src_root, self)? {
+            meta
+        } else {
+            anyhow::bail!("No update metadata for component {} found", self.name());
+        };
+
+        let grub_install = Path::new("/").join(GRUB_BIN);
+        if !grub_install.exists() {
+            bail!("Failed to find {:?}", grub_install);
+        }
+
+        let boot_dir = Path::new(dest_root).join("boot");
+        
+        #[cfg(target_arch = "x86_64")]
+        let r = std::process::Command::new(grub_install)
+            .args(&["--target", "i386-pc"])
+            .args(&["--boot-directory", boot_dir.to_str().unwrap()])
+            .args(&["--modules", "mdraid1x"])
+            .arg(&device)
+            .status()?;
+        
+        #[cfg(target_arch = "powerpc64")]
+        let r = std::process::Command::new(grub_install)
+            .args(&["--target", "powerpc-ieee1275"])
+            .args(&["--boot-directory", boot_dir.to_str().unwrap()])
+            .arg("--no-nvram")
+            .arg(&device)
+            .status()?;
+
+        if !r.success() {
+            anyhow::bail!("Failed to run grub2-install");
+        }
+        Ok(InstalledContent {
+            meta,
+            filetree: None,
+            adopted_from: None,
+        })
+    }
+
+    fn generate_update_metadata(&self, sysroot_path: &str) -> Result<ContentMetadata> {
+        let grub_install = Path::new(sysroot_path).join(GRUB_BIN);
+        if !grub_install.exists() {
+            bail!("Failed to find {:?}", grub_install);
+        }
+        
+        // Query the rpm database and list the package and build times for all the
+        // files in the EFI system partition. If any files are not owned it is considered
+        // and error condition.
+        let rpmout = {
```

Comment:
Now that I look at this we could probably extract this part to a helper function too.  Do you want to try that too?  Or I can do it if you like.

Reply:
SGTM!